File: src/main/java/core/procedures/characters/CharProc.java
package core.procedures.characters;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.function.Function;

public final class CharProc extends AFn {

  public static final CharProc CHAR_TO_INTEGER = new CharProc("char->integer", ch -> (long)ch);
  public static final CharProc CHAR_UPCASE     = new CharProc("char-upcase", Character::toUpperCase);
  public static final CharProc CHAR_DOWNCASE   = new CharProc("char-downcase", Character::toLowerCase);

  private final String name;
  private final Function<Character, Object> function;

  private CharProc(String name, Function<Character, Object> function) {
    super(new FnArgsBuilder().minArgs(1).maxArgs(1).mandatoryArgsTypes(new Class[]{Character.class}));
    this.name = name;
    this.function = function;
  }

  @Override
  public Object apply1(Object arg) {
    return function.apply((Character) arg);
  }

  @Override
  public String getName() {
    return name;
  }
}

File: src/main/java/core/procedures/characters/CharProc.kt
package core.procedures.characters

import core.procedures.AFn
import core.procedures.Arity.Exactly

class CharProc private constructor(override val name: String, private val function: (Char) -> Any) :
        AFn<Char?, Any?>(arity = Exactly(1), mandatoryArgsTypes = arrayOf(Char::class.javaObjectType)) {

    companion object {
        val CHAR_TO_INTEGER = CharProc("char->integer", Char::toLong)
        val CHAR_UPCASE     = CharProc("char-upcase",   Char::toUpperCase)
        val CHAR_DOWNCASE   = CharProc("char-downcase", Char::toLowerCase)
    }

    override operator fun invoke(arg: Char?) = function(arg!!)
}
